### PR TITLE
internal/docstore/dynamodocstore: handle set types

### LINKED ISSUE #1583: docstore/dynamodocstore: support set types
I noticed that the `dynamodbattribute` package's [Marshal](https://godoc.org/github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute#Marshal) method will encode a `[][]byte` as a "binary data set". We don't support DynamoDB's set types in docstore.

If we natively (meaning, with the `dynamoattribute` package) encode a `[][]byte` and try to decode it with the docstore codec, I believe we'll currently get an error. We could fix that with an `AsSpecial` method that recognizes this case.

If we docstore-encode a `[][]byte`, it will currently encode as a list of of `[]byte`s. I'm not sure how `dynamodbattribute.Unmarshal` will handle decoding that into a `[][]byte`.  But in any case the docstore codec should have an `EncodeSpecial` method that handles this.

### PR DESCRIPTION
**Pull request by shantuo**
The decoder will fail on stringset, numberset and binaryset attributevalues. The encoder will not encode [][]byte types, instead it throws an error and provide a workaround.

Fixes #1583 
Updates #1690

### CODE REVIEW

File: internal/docstore/dynamodocstore/codec.go [hunk 1]
```diff
@@ -332,14 +341,28 @@ func toGoValue(av *dyn.AttributeValue) (interface{}, error) {
...
 	switch v.Type() {
 	case typeOfGoTime:
 		if d.av.S == nil {
 			return false, nil, errors.New("expected string field for time.Time")
 		}
 		t, err := time.Parse(time.RFC3339Nano, *d.av.S)
 		return true, t, err
-	default:
-		return false, nil, nil
+	case typeOfBinarySet:
```
Comment (jba) [inline internal/docstore/dynamodocstore/codec.go]: I think what you want to do here is see if `d.av` is one of SS, NS or BS, and return an error. It shouldn't matter what `v` is.

File: internal/docstore/dynamodocstore/codec.go [hunk 2]
```diff
@@ -332,14 +341,28 @@ func toGoValue(av *dyn.AttributeValue) (interface{}, error) {
...
+		return true, nil, fmt.Errorf(unsupportedFmt, `dynamodbav:",binaryset"`)
```
Comment (jba) [inline internal/docstore/dynamodocstore/codec.go]: Instead of writing the struct tag, I would just call it "binary set", etc.

Comment (shantuo) [reply internal/docstore/dynamodocstore/codec.go]: Done.

File: internal/docstore/dynamodocstore/codec.go (hunk 1 above)
Comment (shantuo) [reply internal/docstore/dynamodocstore/codec.go]: Done.

File: internal/docstore/dynamodocstore/codec.go [hunk 3]
```diff
@@ -65,14 +65,23 @@ func (e *encoder) EncodeMap(n int) driver.Encoder {
...
+)
 
 // EncodeSpecial encodes time.Time specially.
 func (e *encoder) EncodeSpecial(v reflect.Value) (bool, error) {
 	switch v.Type() {
 	case typeOfGoTime:
 		ts := v.Interface().(time.Time).Format(time.RFC3339Nano)
 		e.EncodeString(ts)
+	case typeOfBinarySet:
+		return true, errors.New(`unsupported type: [][]byte, the docstore driver
```
[the remark this answers is not in the document]
Comment (shantuo) [reply internal/docstore/dynamodocstore/codec.go]: You're right, there is no data corruption, but we cannot test the native encoder+docstore decoder for `[][]byte` type then.

[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/codec.go]: Right, and that's handled by the `UnsupportedType` feature of the test. If you leave this case here, though, then no one can encode a `[][]byte` at all, which is not what we want.

[the remark this answers is not in the document]
Comment (shantuo) [reply internal/docstore/dynamodocstore/codec.go]: Done.

**Pull request merged**